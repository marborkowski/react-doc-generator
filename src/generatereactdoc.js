import path from "path";
import { parse, resolver } from "react-docgen";
import { promisify } from "util";
import { readFiles } from "node-dir";

import Colors from "colors";
import pkg from "../package.json";

const readFilesPromise = promisify(readFiles);

const templateData = {
  files: [],
  version: pkg.version
};

export const isDefaultValueTypeString = prop => {
  if (!prop || !prop.type) {
    return null;
  }
  return (
    prop.type.name === "string" && typeof prop.defaultValue.value === "string"
  );
};
export const isInvalidDefaultValue = value =>
  /[^\w\s.&:\-+*,!@%$]+/gim.test(value);
export const getTypeOfProp = (prop) => {
  if(!prop){
    return ''
  }
  if(prop.type){
    return prop.type
  }else if(prop.flowType){
    const typeName = prop.flowType.raw ? prop.flowType.raw : prop.flowType.name
    return {
      name: typeName
    }
  }

}  
export function processProp(prop) {
  const { defaultValue = {} } = prop;
  const isString = isDefaultValueTypeString(prop);
  const isInvalidValue = isInvalidDefaultValue(defaultValue.value);
  const processedType = getTypeOfProp(prop)
  const processedDefaultValue =
    defaultValue && isInvalidValue && isString === false
      ? "See code"
      : defaultValue.value;
  const processedDescription = prop.description
    ? prop.description
        .split("\n")
        .map(text => text.replace(/(^\s+|\s+$)/, ""))
        .map(hasValidValue => hasValidValue)
        .join(" ")
    : "";
    
  return {
    ...prop,
    defaultValue: { ...prop.defaultValue, value: processedDefaultValue },
    description: processedDescription,
    type: processedType,
  };
}
function parseSingleFile(fileContent) {
  const components = parse(
    fileContent,
    resolver.findAllExportedComponentDefinitions
  );
  const componentObjects = components.map(component => {
    const { description, displayName } = component;
    const modifiedTitle =
      description && !displayName
        ? description.match(/^(.*)$/m)[0]
        : displayName;
    let modifiedDescription = null;
    if (description) {
      if (description.split("\n").length > 1) {
        modifiedDescription = description.replace(/[\w\W]+?\n+?/, "");
        modifiedDescription = modifiedDescription.replace(/(\n)/gm, "   \n");
      }
      modifiedDescription = `${modifiedDescription}   \n\n`;
    }
    // validate default values
    const propEntries = Object.entries(component.props);
    const modifiedPropEntries = propEntries.map(([propName,propObject]) => {
        const modifiedProp = processProp(propObject)
        return [propName,modifiedProp]   
    })
    const modife = modifiedPropEntries.reduce((accum,current) => {
        const modifiedAccum = {...accum, [current[0]]: current[1]}
        return modifiedAccum
    },{})
    return {
      ...component,
      title: modifiedTitle,
      description: modifiedDescription,
      props: modife
    };
  });
  return componentObjects;
}
async function generateReactDocs({
  sourceDir,
  extensions = [],
  excludePatterns = [],
  ignoreDirectory = [],
  outputDir
}) {
  const cliOutput = [];
  const inputPath = path.resolve(sourceDir);
  await readFilesPromise(
    inputPath,
    {
      match: new RegExp("\\.(?:" + extensions.join("|") + ")$"),
      exclude: excludePatterns,
      excludeDir: ignoreDirectory
    },
    (err, content, filename, next) => {
      if (err) {
        console.log(err, "error");
        throw err;
      }
      try {
        const components = parseSingleFile(content);
        templateData.files.push({ filename, components });
        cliOutput.push([filename, components.length, Colors.green(`OK.`)]);
      } catch (e) {
        console.error("In error",e);
        cliOutput.push([
          filename,
          0,
          Colors.red(`You have to export at least one valid React Class!`)
        ]);
      }

      next();
    }
  );
  if (templateData.files.length === 0) {
    let allExtensions = extensions.map(ext => {
      return `\`*.${ext}\``;
    });
    console.log(
      `${Colors.bold.yellow("Warning:")} ${Colors.yellow(
        `Could not find any files matching the file type: ${allExtensions.join(
          " OR "
        )}`
      )}\n`
    );
  } else {
    console.log(`${cliOutput.toString()}\n\n`);
  }

  return [templateData,cliOutput];
}
export default generateReactDocs;
