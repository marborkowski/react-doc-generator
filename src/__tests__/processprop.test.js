import { processProp } from '../generatereactdoc'

describe('ProcessProp', () => {
    it('Process prop should not modify the default value if the type is a string', () => {
        const mockProp = {
            type: {
                name:'string'
            },
            defaultValue: {
                value:'somestring'
            }
        }
        const output = processProp(mockProp)
        expect(output).toMatchSnapshot()
    })
    it('Process prop should not modify the default value if the type is not string but default value doesnt have special characters', () => {
        const mockProp = {
            type: {
                name: 'number',
            },
            defaultValue: {
                value:22
            }
        }
        const output = processProp(mockProp)
        expect(output).toMatchSnapshot()
    })
    it('Process prop should overwrite the default value when type is not string and there are invalid characters', () => {
        const mockProp = {
            type: {
                name: 'number',
            },
            defaultValue: {
                value:'@#454'
            }
        }
        const output = processProp(mockProp)
        expect(output).toMatchSnapshot()
    })
})