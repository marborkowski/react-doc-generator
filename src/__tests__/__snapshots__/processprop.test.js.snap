// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessProp Process prop should not modify the default value if the type is a string 1`] = `
Object {
  "defaultValue": Object {
    "value": "somestring",
  },
  "description": "",
  "type": Object {
    "name": "string",
  },
}
`;

exports[`ProcessProp Process prop should not modify the default value if the type is not string but default value doesnt have special characters 1`] = `
Object {
  "defaultValue": Object {
    "value": 22,
  },
  "description": "",
  "type": Object {
    "name": "number",
  },
}
`;

exports[`ProcessProp Process prop should overwrite the default value when type is not string and there are invalid characters 1`] = `
Object {
  "defaultValue": Object {
    "value": "See code",
  },
  "description": "",
  "type": Object {
    "name": "number",
  },
}
`;
